telegram/_bot.py

```diff
@@ -188,6 +195,29 @@ def __init__(
                 private_key, password=private_key_password, backend=default_backend()
             )
 
+    def __eq__(self, other: object) -> bool:
+        return self.bot == other
+
+    def __hash__(self) -> int:
+        return hash(self.bot)
+
+    def __reduce__(self) -> NoReturn:
+        """Called by pickle.dumps(). Serializing bots is unadvisable, so we forbid pickling."""
+        raise pickle.PicklingError('Bot objects cannot be pickled!')
+
+    def __deepcopy__(self, memodict: dict) -> 'Bot':
+        """Bot objects can't be deepcopied since they are not pickable, so just return self."""
+        return self
```

Comment:
I disagree with this. trying to deepcopy a bot instance should either fail or return an actual deep copy. If any of the other changes in this PR rely on this, we'll have to rethink them

Reply:
This only comes into effect when trying to directly deepcopy a `Bot`. So if this is removed, then for e.g. context.user_data['bot_inst'] = Bot(...)` [will fail to save](https://github.com/python-telegram-bot/python-telegram-bot/pull/2893/files#diff-0f06c3fbe6411bfe0c9836f2e4f317ba382b8cac774aa2d1a015b857063ab4d9R421)

If the bot is in a `TelegramObject`, then it is removed, deepcopied and re-inserted back in. And thus removing this won't be an issue.

So we can either document saying that bots won't actually be deepcopied, or we remove this method and say that directly serializing bots is not possible

Reply:
good, then let's remove. documenting something about this is maybe not strictly necessary - deepcopying bots directly is probably an even more rare use case than pickling them directly. BTW: I guess the fact thet `deepcopy` and `pickle` use the same interface should be considered more an implementation detail than intended behavior, i.e. I'd always try to make a clear distinction between serializing in general, picking as specific serialization method & copying via `deepcopy`.